xarray/core/dataarray.py

```diff
@@ -158,6 +158,28 @@ def _infer_coords_and_dims(
     return new_coords, dims
 
 
+def _check_data_shape(data, coords, dims):
+    if data is dtypes.NA:
+        data = np.nan
+    if (
+        coords is not None
+        and utils.is_scalar(data, include_0d=False)
+        and not isinstance(data, indexing.ExplicitlyIndexed)
+    ):
+        if utils.is_dict_like(coords):
+            if dims is None:
+                return data
+            else:
+                data_shape = tuple(
+                    as_variable(coords[k], k).size if k in coords.keys() else 1
+                    for k in dims
+                )
+        else:
+            data_shape = tuple(as_variable(coord, "foo").size for coord in coords)
```

Comment:
I had to pass a dummy name to `as_variable` to make it return an object on which I can call `size`, otherwise it wouldn't work if `coords` was passed as a list of `ndarray`.

Reply:
Yeah, this is a bit hacky, but it works. 

Though would `as_compatible_data` work without the `"foo"`? Not everything it returns will have a `.size`, but potentially everything we want?

Reply:
I think this is fine. The "proper" way to do this might be to use matching elements from `dims`, e.g., `tuple(as_variable(coord, dim).size for coord, dim in zip(coords, dims))`. But it's better to use a dummy value, because error checking on `dims` doesn't happen until `_infer_coords_and_dims`.